test/HTML5/Parser/TokenizerTest.php

```diff
@@ -526,7 +526,7 @@ public function testTagNotClosedAfterTagName()
         $this->assertEventError($events->get(0));
         $this->assertEventEquals('startTag', 'span', $events->get(1));
         $this->assertEventError($events->get(2));
-        $this->assertEventEquals('text', '>02', $events->get(3));
+        $this->assertEventEquals('text', '<>02', $events->get(3));
```

Comment:
In that case, the parser would be in the [13.2.5.8 Tag name state](https://html.spec.whatwg.org/multipage/parsing.html#tag-name-state) - thus, form my POV the startTag should be `span<` and the text just `02`.

`\Masterminds\HTML5\Parser\Tokenizer::consumeData` checks for `is_alpha($tok)` and thus skips `<` for the tag name. It seems the states are "optimized" in the current implementation.

In general I think that the current state of the test is still fine (having `<>02` as `text`).
There might be a comment above the test that there is this divergence to the specs.

Reply:
Yes, the parser _should_ be in tag name state, but tag name parsing does not quite follow the spec. It only consumes a [restricted set of characters](https://github.com/Masterminds/html5-php/blob/7a577d6651727f04d4dc145ac2995e9b9419e7cf/src/HTML5/Parser/Tokenizer.php#L385) and then enters attribute state, where the spec indicates that the character should be part of the tag name. Attribute state parsing also does not quite follow the spec in that it [exits as soon as it encounters the `<`](https://github.com/Masterminds/html5-php/blob/7a577d6651727f04d4dc145ac2995e9b9419e7cf/src/HTML5/Parser/Tokenizer.php#L468-L474), where the spec indicates it should be [treated as part of an attribute name while in that state](https://html.spec.whatwg.org/multipage/parsing.html#attribute-name-state) (thus `<span <>02</span>` also does not produce the correct output). At this point the parser re-enters the data state and the character combo `<>` does not match a known state and so is consumed as data. With the change to the data state handling the extra `<` is exposed whereas before it was just swallowed.

That was the meaning behind my [comment in 250](https://github.com/Masterminds/html5-php/issues/250#issuecomment-2272578028) that there are other issues to address. It took me a while to get the data state parsing right, so I didn't want to immediately dive into the parsing issues with the other states.

Reply:
The original issue is referencing problems in the data state and that was what I was attempting to address. Instead of a comment here, the parsing issues in tag name and attribute name could be addressed now in this PR _or_ a new issue outlining the problem in those states could be opened with information about the divergence. I was primarily waiting to see what feedback I got on the current change before deciding whether to address those issues here or open a new issue and defer the additional work.